## 1. Preprocessing
- Remove URLs, HTML tags
- Standardizing
- Handle irrelevant characters
- Tokenization
- Lowercasing
- Remove stop words or rare words
- Spelling Correction
- Stemming and Lemmatization

In [2]:
import re
import string

In [3]:
def preprocessing(line):
    # Lowercasing
    line = line.lower()
    # Remove punctuation
    translation_table = str.maketrans(" ", " ", string.punctuation)
    line = line.translate(translation_table)
    line = line.replace("“", " ")
    line = line.replace("”", " ")
    # Remove number
    line = re.sub(r'\d+', '', line)
    # Remove blank
    line = line.strip()
    while line.find("  ") != -1:
        line = line.replace("  ", " ")
    return line

In [4]:
def write_file(type):
    path_raw = f"./raw_data/{type}.txt"
    with open(path_raw, "r", encoding="utf-8") as f_raw:
        lines = f_raw.readlines()

    path_process = f"./process_data/{type}.txt"
    with open(path_process, "w", encoding="utf-8") as f_process:
        for line in lines:
            line = preprocessing(line)
            if line != None and len(line) != 0:
                f_process.write(line + '\n')

In [5]:
types = ["vanhoa", "phapluat", "thethao", "giaoduc", "chinhtri", "kinhte", "xahoi"]
for type in types:
    write_file(type)

#### => Result
#### Before: Trao 250 suất học bổng chương trình "Vì em hiếu học" năm 2022 tại Sóc Trăng
#### After: trao suất học bổng chương trình vì em hiếu học năm tại sóc trăng
<hr>

## 2. EDA

In [6]:
words = set()
for type in types:
    with open(f"./process_data/{type}.txt", "r", encoding="utf-8") as f:
        for line in f.readlines():
            max_len = max(max_len, len(line))
            line = line.rstrip("\n")
            chars = line.split(" ")
            for char in chars:
                words.add(char)
print(len(words))

4456


#### => Tổng số từ sử dụng trong dữ liệu là: 4456 từ

In [12]:
def cal_num_word(line):
    num = 0
    while line.find(" ") != -1:
        num += 1
        line = line.replace(" ", "", 1)
    return num + 1

max_len = 0
for type in types:
    with open(f"./process_data/{type}.txt", "r", encoding="utf-8") as f:
        for line in f.readlines():
            line = line.rstrip("\n")
            if max_len < cal_num_word(line):
                max_len = cal_num_word(line)

print(max_len)

49


#### => Câu dài nhất có số từ là: 49 từ